android/app/src/main/java/com/zulipmobile/notifications/FcmMessage.kt

```diff
@@ -187,19 +187,38 @@ internal data class MessageFcmMessage(
     }
 }
 
-internal data class RemoveFcmMessage(
+data class RemoveFcmMessage(
+    val identity: Identity?,
     val messageIds: Set<Int>
 ) : FcmMessage() {
     companion object {
         fun fromFcmData(data: Map<String, String>): RemoveFcmMessage {
             val messageIds = HashSet<Int>()
+            val identity = data["server"]?.let { serverHost ->
+                Identity(
+                    // `server` was added in server version 1.8.0
+                    // (released 2018-04-16; commit 014900c2e).
+                    serverHost = serverHost,
+
+                    // `realm_id` was added in the same commit as `server`.
+                    realmId = data.require("realm_id").parseInt("realm_id"),
+
+                    // `realm_uri` was added in server version 1.9.0
+                    // (released 2018-11-06; commit 5f8d193bb).
+                    realmUri = data["realm_uri"]?.parseUrl("realm_uri"),
+
+                    // `user` was already present in server version 1.6.0 .
+                    email = "" //TODO
```

Comment:
Remove FCM message does't have this data (`user` i.e`email`) so how we are going to handle this.

With the new data structure of the conversation map where key is the combination of `Identity` and `zulipMessageId`. Now how we are going to compare the key (especially, `Identity`) when we receive remove FCM message event, so that we can remove it from conversation map.

cc @gnprice

Reply:
Yeah, this is something we should fix on the server. :-/ But of course we still have to handle the case where it's missing.

To do that, we can just iterate through all the keys in the map. (The top-level keys should be just `Identity` -- see more detailed comment I just made in this PR's main thread.) There won't be a ton of them. When the `RemoveFcmMessage` lacks an `email`, we'll just have to apply it to *every* identity that has the right server and realm. For almost all normal users, there will be just one of those anyway.

Reply:
Also, maybe you were planning to change this anyway (I know this is a WIP commit), but to be explicit:
* Let's never use an empty string (or zero, or empty list/array, etc.) as a symbol for "don't know". This is a notorious source of bugs. Instead, use an explicit "no data" marker -- in Kotlin or Java, that's `null`.
* This means you'll have to change the types! The type of `Identity#email` will go from `String` to `String?`. The fact you have to change this type is almost the same as the whole point :slightly_smiling_face: -- it means everywhere we use this field, we'll have to confront the possibility that it's `null`. Which is a good thing, because that is a real possibility!

Reply:
cool! Submitted https://github.com/zulip/zulip/issues/11961 to track this on server. :)